src/reply_parser.c

```diff
@@ -0,0 +1,192 @@
+#include "reply_parser.h"
+#include "server.h"
+
+/* ----------------------------------------------------------------------------------------
+ * A resp parser used to parse replies returned by RM_Call
+ * or Lua redis.call. The parser introduce callbacks
+ * that needs to be set by the user. Each callback represents
+ * a different reply type. Each callback gets a p_ctx that
+ * was given to the parseReply function. The callbacks also give the
+ * protocol (underline blob) of the current reply and the size.
+ *
+ * 3 special callbacks also get the parser object itself
+ * (array_callback, set_callback, map_callback). This is because
+ * it's their responsibility to continue the parsing by calling
+ * parseReply. When the user continues parsing he can give a
+ * different p_ctx (this is used by CallReply to give the nested reply as the p_ctx).
+ * Also notice that those 3 callbacks do not get the proto len, this is because
+ * When calling those callbacks the proto len is still unknown. the user
+ * can calculate the len after finish parsing the entire collection.
+ * ---------------------------------------------------------------------------------------- */
+
+static int parseBulk(ReplyParser* parser, void* p_ctx) {
+    const char *proto = parser->curr_location;
+    char *p = strchr(proto+1,'\r');
+    long long bulklen;
+    parser->curr_location = p + 2; /* for \r\n */
+
+    string2ll(proto+1,p-proto-1,&bulklen);
+    if (bulklen == -1) {
+        parser->null_bulk_string_callback(p_ctx, proto, parser->curr_location - proto);
+    } else {
+        const char* str = parser->curr_location;
+        parser->curr_location += bulklen;
+        parser->curr_location += 2; /* for \r\n */
+        parser->bulk_string_callback(p_ctx, str, bulklen, proto, parser->curr_location - proto);
+    }
+
+    return C_OK;
+}
+
+static int parseSimpleString(ReplyParser* parser, void* p_ctx) {
+    const char *proto = parser->curr_location;
+    char *p = strchr(proto+1,'\r');
```

Comment:
Handle `p==NULL` here as well? This repeats itself.

Reply:
Discussion summary:
we will leave error handling for another PR (because the old parser did not check for protocol correctness as well, so we are not changing the behavior). On the other PR we should add error handling code and make sure we do not see any performance regression. I guess this covers the other comments about error checking above, @yossigo please confirm.